src/Client/ClientBase.cpp

```diff
@@ -74,6 +74,21 @@ namespace ErrorCodes
 namespace DB
 {
 
+volatile sig_atomic_t need_cancel_query = false;
+volatile sig_atomic_t query_cancelled = true;
+
+/// This signal handler is set only for sigint.
+void signalHandler(int signum)
+{
+    if (query_cancelled)
+        return;
+    /// Exit if it is a second sigint and query is not cancelled yet.
+    if (need_cancel_query && !query_cancelled)
```

Comment:
Two independent flags looks too complex and suspicious, because it's hard to understand what will happen in all possible execution orders. Maybe it's possible to use one atomic counter or something like that?